### PR TITLE
Design Redux store to store progress of data processing

### PR DESCRIPTION
**Pull request by marcellp**
# Background
#### Link to issue
https://biomage.atlassian.net/browse/BIOMAGE-501

#### Link to staging deployment URL
https://ui-krdmkow7rk4jjph7kr83qpeawr.scp-staging.biomage.net/

#### Links to any Pull Requests related to this
N/A

#### Anything else the reviewers should know about the changes here
Also fixed some additional error messages that kept cropping up in the data processing page, as they made debugging this kind of hard.

Added some UX improvements that will mark the steps completed, currently being performed, and still outstanding using colors as well as icons for accessibility. Vicky should take a look and evaluate if she is satisfied.

# Changes
### Code changes
Modified the old Redux store to keep track of state of data processing. The data processing page is also reflecting the changes.

# Definition of DONE
Your changes will be ready for merging after each of the steps below have been completed:

### Testing
- [x] Unit tests written
- [x] Tested locally with Inframock
- [x] Deployed to staging

To set up easy local testing with inframock, follow the instructions here: https://github.com/biomage-ltd/inframock
To deploy to the staging environment, follow the instructions here: https://github.com/biomage-ltd/biomage-utils

### Documentation updates
Is all relevant documentation updated to reflect the proposed changes in this PR?

- [x] Relevant Github READMEs updated
- [x] Relevant wiki pages created/updated

### Approvers
- [ ] Approved by a member of the core engineering team
- [ ] (UX changes) Approved by @vickymorrison

### Just before merging:
- [ ] After the PR is approved, the `unstage` script in here: https://github.com/biomage-ltd/biomage-utils is executed. This script cleans up your deployment to staging

### Optional
- [x] Photo of a cute animal attached to this PR
image

### CODE REVIEW

File: src/pages/experiments/[experimentId]/data-processing/data-integration/components/CalculationConfig.jsx [hunk 1]
```diff
@@ -61,11 +61,20 @@ const CalculationConfig = () => {
...
       <Form size='small'>
+        {changesOutstanding && (
+          <Form.Item>
+            <Alert
+              message='Your changes are not yet applied. To update the plots, click Apply.'
+              type='warning'
+              showIcon
+            />
+          </Form.Item>
+        )}
```
Comment (marcellp) [inline src/pages/experiments/[experimentId]/data-processing/data-integration/components/CalculationConfig.jsx:105]: These changes were made to bring Data Integration more in line with what we have now with Configure Embedding.

File: src/pages/experiments/[experimentId]/data-processing/data-integration/index.jsx [hunk 2]
```diff
@@ -22,10 +22,10 @@ class ProcessingViewPage extends React.Component {
           extra={<FeedbackButton />}
         />
 
-        <Collapse accordion defaultActiveKey={['1']}>
+        <Collapse accordion defaultActiveKey={['dim-red']}>
           <Panel
             header='Dimensionality Reduction'
-            key='1'
+            key='dim-red'
```
Comment (marcellp) [inline src/pages/experiments/[experimentId]/data-processing/data-integration/index.jsx:28]: Fixing some errors that caused warnings in the console.

File: src/pages/experiments/[experimentId]/data-processing/filter-cells/components/CellSizeDistribution/CellSizeDistribution.jsx [hunk 3]
```diff
@@ -507,19 +506,19 @@ class CellSizeDistribution extends React.Component {
...
-                <Form.Item label="Min cell size:">
+            <Space direction='vertical' style={{ width: '100%' }} />
+            <Collapse defaultActiveKey={['filtering-settings']}>
+              <Panel header='Filtering Settings' collapsible={!filtering ? 'disabled' : 'header'} key='filtering-settings'>
+                <Form.Item label='Min cell size:'>
                   <InputNumber
-                    disabled={!filtering}
-                    onPressEnter={val => changeCellSize(val)}
+                    collapsible={!filtering ? 'disabled' : 'header'}
+                    onPressEnter={(val) => changeCellSize(val)}
```
Comment (marcellp) [inline src/pages/experiments/[experimentId]/data-processing/filter-cells/components/CellSizeDistribution/CellSizeDistribution.jsx:521]: Similarly, fixes that caused warnings in console.

File: src/pages/experiments/[experimentId]/data-processing/index.jsx [hunk 4]
```diff
@@ -101,18 +102,47 @@ const DataProcessingPage = () => {
...
+                  )}
+
+                  {!completedSteps.has(key) && stepId !== i && (
+                    <Text
+                      disabled
+                    >
+                      <EllipsisOutlined />
+                      <span style={{ marginLeft: '0.25rem' }}>{name}</span>
+                    </Text>
+                  )}
```
Comment (marcellp) [inline src/pages/experiments/[experimentId]/data-processing/index.jsx]: Some UX changes that make it easier to navigate the dropdown menu regarding which steps are done, outstanding, currently filling.

File: src/pages/experiments/[experimentId]/data-processing/filter-cells/components/GenesVsUMIs/GenesVsUMIs.jsx [hunk 5]
```diff
@@ -171,7 +171,7 @@ class GenesVsUMIs extends React.Component {
                 x2: {
                   scale: 'xscale',
                   field: 'bin1',
-                  //offset: { signal: 'binStep > 0.02 ? -0.5 : 0' },
+                  // offset: { signal: 'binStep > 0.02 ? -0.5 : 0' },
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/data-processing/filter-cells/components/GenesVsUMIs/GenesVsUMIs.jsx]: just remove this if it is not used

File: src/pages/experiments/[experimentId]/data-processing/index.jsx [hunk 6]
```diff
@@ -101,18 +102,47 @@ const DataProcessingPage = () => {
       <Col span='8'>
         <Select
           value={stepId}
-          onChange={(id) => setStepId(id)}
+          onChange={(id) => {
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/data-processing/index.jsx]: if you are renaming `id` to `key` in the `steps` array above, it makes sense to rename these to `key` too

File: src/pages/experiments/[experimentId]/data-processing/index.jsx [hunk 7]
```diff
@@ -141,11 +171,24 @@ const DataProcessingPage = () => {
                 () => {
                   const newId = Math.min(stepId + 1, steps.length - 1);
```
Comment (ivababukova) [inline src/pages/experiments/[experimentId]/data-processing/index.jsx]: I see, so you are treating the order of the steps as their ids, so for example, cell size distribution has id = 1. I wonder how can we make this more obvious in the code

File: src/pages/experiments/[experimentId]/data-processing/index.jsx (hunk 6 above)
Comment (ivababukova) [reply src/pages/experiments/[experimentId]/data-processing/index.jsx]: so this is actually not the `key`, but the step number, where the number represents the ordering of execution for a step. We definitely need to rename this to make it more obvious

File: src/pages/experiments/[experimentId]/data-processing/index.jsx (hunk 7 above)
Comment (marcellp) [reply src/pages/experiments/[experimentId]/data-processing/index.jsx]: Correct, maybe `index` or `stepIdx` would be better?

Comment (ivababukova) [reply src/pages/experiments/[experimentId]/data-processing/index.jsx]: stepIdx sounds good yes

**Comment by marcellp** [comment]
Deployed to https://ui-krdmkow7rk4jjph7kr83qpeawr.scp-staging.biomage.net/.

**Comment by vickymorrison** [comment]
A couple of comments from the UX side:
1. Can we keep the green tick that marks each step as done, but change the text to black? The green text looks really ugly.
2. In the dropdown menu, surely the user should be able to select the next filter i.e. so that they can navigate to the next filter either using the dropdown or using the 'next' button.

**Comment by marcellp** [comment]
Thanks for the feedback Vicky, fixed both. They are pushed, so should appear soon in staging.

File: src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx [hunk 8]
```diff
@@ -71,34 +71,39 @@ const EmbeddingContinuousPlot = () => {
...
     updatePlotWithChanges({ shownGene: highestDispersionGene });
   }
 
   useEffect(() => {
-    if (!data && processingSettings.configureEmbedding) {
+    if (!data && processingSettings.configureEmbedding && experimentId) {
       dispatch(loadEmbedding(experimentId, embeddingType));
     }
-  }, [processingSettings]);
+  }, [processingSettings, experimentId]);
```
Comment (marcellp) [inline src/pages/experiments/[experimentId]/plots-and-tables/embedding-continuous/index.jsx:106]: Fixes a bug we had in plots and tables that I noticed.

File: src/pages/experiments/[experimentId]/plots-and-tables/embedding-categorical/index.jsx [hunk 9]
```diff
@@ -78,10 +78,10 @@ const EmbeddingCategoricalPlot = () => {
   }, [experimentId]);
 
   useEffect(() => {
-    if (!data && processingSettings.configureEmbedding) {
+    if (!data && processingSettings.configureEmbedding && experimentId) {
       dispatch(loadEmbedding(experimentId, embeddingType));
     }
-  }, [processingSettings]);
+  }, [processingSettings, experimentId]);
```
Comment (marcellp) [inline src/pages/experiments/[experimentId]/plots-and-tables/embedding-categorical/index.jsx:84]: Similarly.

File: src/components/ContentWrapper.jsx [hunk 10]
```diff
@@ -135,22 +154,25 @@ const ContentWrapper = (props) => {
...
+            }
+            mode='inline'
+          >
+            {menuLinks.map(({ path, icon, name }) => (
+              <Menu.Item key={path} icon={icon}>
+                <Link as={path.replace('[experimentId]', experimentId)} href={path} passHref>
+                  <a>{name}</a>
+                </Link>
+              </Menu.Item>
+            ))}
```
Comment (marcellp) [inline src/components/ContentWrapper.jsx]: Refactored as a prep for the ticket regarding navigation.

**Review by vickymorrison — APPROVED** [review]

**Review by ivababukova — APPROVED** [review]

**Pull request merged**